File: dotnet/SketchOxide/src/Frequency/ElasticSketch.cs
using System;
using System.Text;
using SketchOxide.Native;

namespace SketchOxide.Frequency;

/// <summary>
/// Elastic Sketch for adaptive frequency estimation.
///
/// A two-layer sketch that adapts to the skewness of the data distribution.
/// Heavy hitters are stored in a hash table (heavy part) while the light part
/// uses Count-Min style counters. This provides better accuracy for heavy hitters
/// while maintaining efficiency for the long tail.
/// </summary>
/// <remarks>
/// Note: ElasticSketch does not support merging as the adaptive state cannot be
/// meaningfully combined across multiple instances.
/// </remarks>
public sealed class ElasticSketch : NativeSketch
{
    private readonly uint _bucketCount;
    private readonly uint _depth;

    /// <summary>
    /// Creates a new Elastic Sketch with the specified parameters.
    /// </summary>
    /// <param name="bucketCount">Number of buckets in the heavy part. Must be greater than 0.</param>
    /// <param name="depth">Depth of the light part (number of hash functions). Must be greater than 0.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if bucketCount or depth is 0.</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public ElasticSketch(uint bucketCount, uint depth)
    {
        if (bucketCount == 0)
            throw new ArgumentOutOfRangeException(nameof(bucketCount), bucketCount, "Bucket count must be greater than 0");
        if (depth == 0)
            throw new ArgumentOutOfRangeException(nameof(depth), depth, "Depth must be greater than 0");

        _bucketCount = bucketCount;
        _depth = depth;
        NativePtr = SketchOxideNative.elasticsketch_new(bucketCount, depth);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native ElasticSketch");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private ElasticSketch(uint bucketCount, uint depth, nuint ptr)
    {
        _bucketCount = bucketCount;
        _depth = depth;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the number of buckets in the heavy part.
    /// </summary>
    public uint BucketCount
    {
        get
        {
            CheckAlive();
            return _bucketCount;
        }
    }

    /// <summary>
    /// Gets the depth of the light part.
    /// </summary>
    public uint Depth
    {
        get
        {
            CheckAlive();
            return _depth;
        }
    }

    /// <summary>
    /// Updates the sketch with a new element, incrementing its count.
    /// </summary>
    /// <param name="data">The bytes representing the item to count.</param>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public void Update(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                SketchOxideNative.elasticsketch_update(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Updates the sketch with a string element, incrementing its count.
    /// </summary>
    /// <param name="value">The string item to count.</param>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public void Update(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        Update(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Estimates the frequency count of an element.
    /// </summary>
    /// <param name="data">The bytes representing the item to query.</param>
    /// <returns>The estimated count for the item.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public ulong Estimate(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.elasticsketch_estimate(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Estimates the frequency count of a string element.
    /// </summary>
    /// <param name="value">The string item to query.</param>
    /// <returns>The estimated count for the item.</returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public ulong Estimate(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Estimate(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Serializes the sketch to a byte array.
    /// </summary>
    /// <returns>Serialized sketch bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.elasticsketch_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes an Elastic Sketch from a byte array.
    /// </summary>
    /// <param name="data">Serialized sketch bytes.</param>
    /// <param name="bucketCount">The bucket_count parameter used when creating the original sketch.</param>
    /// <param name="depth">The depth parameter used when creating the original sketch.</param>
    /// <returns>A new ElasticSketch instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static ElasticSketch Deserialize(byte[] data, uint bucketCount, uint depth)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.elasticsketch_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize ElasticSketch: invalid data");

        return new ElasticSketch(bucketCount, depth, ptr);
    }

    /// <summary>
    /// Returns a string representation of the sketch.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "ElasticSketch(disposed)";
        return $"ElasticSketch(bucketCount={_bucketCount}, depth={_depth})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.elasticsketch_free(NativePtr);
            NativePtr = 0;
        }
    }
}

File: java/src/main/java/com/sketches_oxide/frequency/ElasticSketch.java
package com.sketches_oxide.frequency;

import com.sketches_oxide.MergeableSketch;
import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * ElasticSketch - Adaptive Bucket Allocation for Network Monitoring.
 *
 * <p>ElasticSketch is a frequency estimation data structure optimized for network
 * traffic monitoring. It uses adaptive bucket allocation to efficiently handle
 * both heavy hitters and light flows in the same structure.
 *
 * <p><strong>Algorithm Overview:</strong>
 * <p>ElasticSketch combines two components:
 * <ol>
 *   <li><strong>Heavy Part</strong>: A hash table for tracking heavy hitters
 *       with exact counts when possible</li>
 *   <li><strong>Light Part</strong>: A Count-Min Sketch variant for tracking
 *       light flows</li>
 * </ol>
 *
 * <p>Items are dynamically moved between parts based on their frequency,
 * providing accuracy for both common and rare items.
 *
 * <p><strong>Key Features:</strong>
 * <ul>
 *   <li>Adaptive allocation between heavy and light flows</li>
 *   <li>High accuracy for heavy hitters</li>
 *   <li>Memory-efficient for light flows</li>
 *   <li>Optimized for network traffic patterns</li>
 * </ul>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>Network traffic flow monitoring</li>
 *   <li>DDoS attack detection</li>
 *   <li>Load balancing decisions</li>
 *   <li>Network anomaly detection</li>
 *   <li>Per-flow statistics collection</li>
 * </ul>
 *
 * <p><strong>Space Complexity:</strong> O(bucketCount * depth)
 *
 * <p><strong>Example:</strong>
 * <pre>
 * try (ElasticSketch es = new ElasticSketch(4096, 4)) {
 *     // Process network packets
 *     for (byte[] flowKey : packets) {
 *         es.update(flowKey);
 *     }
 *
 *     // Query flow frequency
 *     long count = es.estimate(flowKey);
 *     System.out.println("Flow packets: " + count);
 * }
 * </pre>
 *
 * @see <a href="https://dl.acm.org/doi/10.1145/3230543.3230544">ElasticSketch Paper</a>
 * @see CountMinSketch
 * @see SpaceSaving
 */
public final class ElasticSketch extends NativeSketch implements MergeableSketch<ElasticSketch> {

    /** Minimum allowed bucket count. */
    public static final long MIN_BUCKET_COUNT = 1;

    /** Maximum allowed bucket count. */
    public static final long MAX_BUCKET_COUNT = 1_000_000_000L;

    /** Minimum allowed depth. */
    public static final long MIN_DEPTH = 1;

    /** Maximum allowed depth. */
    public static final long MAX_DEPTH = 32;

    /** Default bucket count. */
    public static final long DEFAULT_BUCKET_COUNT = 4096;

    /** Default depth. */
    public static final long DEFAULT_DEPTH = 4;

    private final long bucketCountParam;
    private final long depthParam;

    /**
     * Create a new ElasticSketch with default parameters.
     *
     * @throws OutOfMemoryException if native memory allocation fails
     */
    public ElasticSketch() {
        this(DEFAULT_BUCKET_COUNT, DEFAULT_DEPTH);
    }

    /**
     * Create a new ElasticSketch.
     *
     * @param bucketCount number of buckets in the heavy part (1 to 1,000,000,000)
     *                    Higher values provide better accuracy but use more memory.
     * @param depth number of hash functions/rows (1 to 32)
     *              Higher values reduce collision errors.
     * @throws IllegalArgumentException if parameters are invalid
     * @throws OutOfMemoryException if native memory allocation fails
     */
    public ElasticSketch(long bucketCount, long depth) {
        if (bucketCount < MIN_BUCKET_COUNT || bucketCount > MAX_BUCKET_COUNT) {
            throw new IllegalArgumentException(
                    "bucketCount must be between " + MIN_BUCKET_COUNT + " and " + MAX_BUCKET_COUNT +
                            ", got: " + bucketCount);
        }
        if (depth < MIN_DEPTH || depth > MAX_DEPTH) {
            throw new IllegalArgumentException(
                    "depth must be between " + MIN_DEPTH + " and " + MAX_DEPTH +
                            ", got: " + depth);
        }

        this.bucketCountParam = bucketCount;
        this.depthParam = depth;
        this.nativePtr = SketchOxideNative.elasticsketch_new(bucketCount, depth);

        if (this.nativePtr == 0) {
            throw new OutOfMemoryException("Failed to allocate ElasticSketch");
        }
    }

    /**
     * Update the frequency of an item (increment by 1).
     *
     * @param item the item to count (typically a flow key)
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public void update(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.elasticsketch_update(nativePtr, item);
    }

    /**
     * Update with a string item.
     *
     * @param item the string to count
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public void update(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        update(item.getBytes());
    }

    /**
     * Estimate the frequency of an item.
     *
     * <p>Returns an estimate that is accurate for heavy hitters and
     * bounded for light flows.
     *
     * @param item the item to query
     * @return estimated frequency
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public long estimate(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return SketchOxideNative.elasticsketch_estimate(nativePtr, item);
    }

    /**
     * Estimate the frequency of a string item.
     *
     * @param item the string to query
     * @return estimated frequency
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public long estimate(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return estimate(item.getBytes());
    }

    /**
     * Merge another ElasticSketch into this one.
     *
     * <p>Both sketches must have the same bucketCount and depth parameters.
     *
     * @param other the sketch to merge
     * @throws IllegalStateException if either sketch is closed
     * @throws IllegalArgumentException if parameters don't match
     * @throws NullPointerException if other is null
     */
    @Override
    public void merge(ElasticSketch other) {
        checkAlive();
        if (other == null) {
            throw new NullPointerException("other cannot be null");
        }
        other.checkAlive();

        if (this.bucketCountParam != other.bucketCountParam || this.depthParam != other.depthParam) {
            throw new IllegalArgumentException(
                    "Cannot merge ElasticSketches with different parameters: " +
                            String.format("(%d, %d) vs (%d, %d)",
                                    this.bucketCountParam, this.depthParam,
                                    other.bucketCountParam, other.depthParam));
        }

        SketchOxideNative.elasticsketch_merge(this.nativePtr, other.nativePtr);
    }

    /**
     * Get the bucket count parameter.
     *
     * @return the number of buckets in the heavy part
     * @throws IllegalStateException if the sketch has been closed
     */
    public long bucketCount() {
        checkAlive();
        return SketchOxideNative.elasticsketch_bucketCount(nativePtr);
    }

    /**
     * Get the depth parameter.
     *
     * @return the number of hash functions/rows
     * @throws IllegalStateException if the sketch has been closed
     */
    public long depth() {
        checkAlive();
        return SketchOxideNative.elasticsketch_depth(nativePtr);
    }

    /**
     * Serialize the sketch to binary format.
     *
     * @return binary representation of the sketch
     * @throws IllegalStateException if the sketch has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.elasticsketch_serialize(nativePtr);
    }

    /**
     * Deserialize from binary format.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new ElasticSketch instance
     * @throws IllegalArgumentException if data is invalid
     * @throws NullPointerException if data is null
     */
    public static ElasticSketch deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.elasticsketch_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized ElasticSketch data");
        }

        ElasticSketch es = new ElasticSketch(DEFAULT_BUCKET_COUNT, DEFAULT_DEPTH);
        SketchOxideNative.elasticsketch_free(es.nativePtr);
        es.nativePtr = ptr;
        return es;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.elasticsketch_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        try {
            if (nativePtr != 0) {
                return String.format("ElasticSketch(bucketCount=%d, depth=%d)",
                        bucketCount(), depth());
            }
        } catch (IllegalStateException e) {
            // Closed
        }
        return "ElasticSketch(closed)";
    }
}
